src/products/components/ProductStocks/components/ProductStocksAssignWarehouses.tsx

```diff
@@ -1,81 +1,143 @@
+import BackButton from "@dashboard/components/BackButton";
+import Checkbox from "@dashboard/components/Checkbox";
+import { InfiniteScroll } from "@dashboard/components/InfiniteScroll";
+import { DashboardModal } from "@dashboard/components/Modal";
+import ResponsiveTable from "@dashboard/components/ResponsiveTable";
+import TableRowLink from "@dashboard/components/TableRowLink";
 import { WarehouseFragment } from "@dashboard/graphql";
-import { Box, Button, Dropdown, List, Spinner, Text } from "@saleor/macaw-ui-next";
-import React, { useEffect } from "react";
-import { FormattedMessage } from "react-intl";
-
-import { messages } from "../messages";
-import { ProductStocksInfinityScrollList } from "./ProductStocksInfinityScrollList";
+import useSearchQuery from "@dashboard/hooks/useSearchQuery";
+import { CircularProgress, TableBody, TableCell, TextField } from "@material-ui/core";
+import { ConfirmButton } from "@saleor/macaw-ui";
+import { Button, Option, sprinkles } from "@saleor/macaw-ui-next";
+import React, { useState } from "react";
+import { useIntl } from "react-intl";
 
 interface ProductStocksAssignWarehousesProps {
   warehousesToAssign: WarehouseFragment[];
   hasMoreWarehouses: boolean;
   loadMoreWarehouses: () => void;
   onWarehouseSelect: (warehouseId: string, warehouseName: string) => void;
+  loading: boolean;
+  searchWarehouses: (query: string) => void;
 }
 
 export const ProductStocksAssignWarehouses = ({
   hasMoreWarehouses,
   loadMoreWarehouses,
   onWarehouseSelect,
   warehousesToAssign,
+  loading,
+  searchWarehouses,
 }: ProductStocksAssignWarehousesProps) => {
-  useEffect(() => {
-    // In case there is no more warehouses to assign, load more warehouses
-    // Example: from first page all warehouse has been already assigned, so we need to load more warehouses to assign
-    if (hasMoreWarehouses && !warehousesToAssign.length) {
-      loadMoreWarehouses();
-    }
-  }, [hasMoreWarehouses, warehousesToAssign.length]);
+  const [warehouses, setWarehouses] = useState<Option[]>([]);
+  const intl = useIntl();
+  const [open, setOpen] = useState(false);
+  const [query, onQueryChange, queryReset] = useSearchQuery(searchWarehouses);
```

Comment:
for the future, I would consider avoiding distributed states for the same thing. If our logic is
- option A: Modal closed, nothing in context
- option B: Modal open, warehouse(s) in context

Then we have finite state machine and option C & D (like open modal without context) are invalid

in such case we can model

```
class ModalOpenState {
  constructor(wh: Warehouse[]) {}

  readonly type = "modal_open"
  wh: Warehouse[]
}

class ModalClosedState {
  type= "modal_closed"
}

type PossibleStates = ModalOpenState | ModalCloseState

const [state, setState] = useState<PossibleStates>(new ModalOpenState([...]))
```

This may be a little more verbose but it's more readalbe in maintainable, also less buggy, especially for complex parts